### DISCUSSION TITLE
Profile Lexicon

### DISCUSSION THREAD

**Discussion by ghost** [lexicon-community/lexicon#9] [Ideas]
## Intent and Rationale
Within a social graph, users have personal information they want to be able to manage and present to others. The most popular Lexicon namespace, `app.bsky`, contains a record type for describing a user's social profile on Bluesky. Other AT Protocol services have adopted a variety of methods for managing user profiles, including:

* Creating their own profile record that stands alone from a user's Bluesky profile
* Importing profile information from a user's Bluesky profile into their own profile record on creation
* Using a user's Bluesky profile for all profile information

## Intended Use
This Lexicon aims to bring consistency to profile behavior by allowing a user to declare their profile information independent of any service, override their generic profile for specific ATProto services, and declare service-specific profile information.

The new record type `community.lexicon.actor.profile` will be defined and primarily defined with rkey `com.atproto` to declare a user's profile to be used in the absence of applicable service-specific profile records. Service-specific profile records should use their service's NSID (the service's homepage as a reverse-URI) as the rkey, and inherit all object properties from the generic profile record Lexicon. Services are free to define and store additional properties in their profiles.

Care should be taken to ensure user interfaces appropriately inform the user of the difference between the generic profile and the service-specific profile, especially when editing the profile information.

### Generic profile (rkey `com.atproto`)
```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "description": "A declaration of an AT Protocol account profile.",
      "key": "literal:com.atproto",
      "record": {
        "type": "object",
        "properties": {
          "displayName": {
            "type": "string",
            "maxGraphemes": 64,
            "maxLength": 640
          },
          "description": {
            "type": "string",
            "description": "Free-form profile description text.",
            "maxGraphemes": 256,
            "maxLength": 2560
          },
          "avatar": {
            "type": "blob",
            "description": "Small image to be displayed next to posts from account. AKA, 'profile picture'",
            "accept": ["image/png", "image/jpeg"],
            "maxSize": 1000000
          },
          "banner": {
            "type": "blob",
            "description": "Larger horizontal image to display behind profile view."
            "accept": ["image/png", "image/jpeg"],
            "maxSize": 1000000
          }
        }
      }
    }
  }
}
```

### Bluesky profile (rkey `app.bsky`)
```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "description": "A declaration of a Bluesky account profile.",
      "key": "literal:app.bsky",
      "record": {
        "type": "object",
        "properties": {
          "joinedViaStarterPack": {
            "type": "ref",
            "ref": "com.atproto.repo.strongRef"
          },
          "pinnedPost": {
            "type": "ref",
            "ref": "com.atproto.repo.strongRef"
          }
        }
      }
    }
  }
}
```

### Smoke Signal profile (rkey `events.smokesignal`)
```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "description": "A declaration of a Smoke Signal account profile.",
      "key": "literal:events.smokesignal",
      "record": {
        "type": "object",
        "properties": {
          "tz": {
            "type": "string",
            "description": "The time zone of the profile (e.g. 'America/New_York')."
          }
        }
      }
    }
  }
}
```

**Comment by ghost** [comment]
By the way, the `description` property should (imo) be made into an object with `text` and `facets` properties to allow for better marking up of tags and mentions in descriptions. I just copied the definition seen here from the Bluesky Lexicon.

**Comment by ghost** [comment]
Here are some examples of records using the proposed Lexicon to help illustrate:
- [Generic profile](https://pdsls.dev/at/did:plc:erqwpimsmwnzohwoiojsy22z/community.lexicon.alpha.actor.profile/com.atproto)
- [Bluesky profile](https://pdsls.dev/at/did:plc:erqwpimsmwnzohwoiojsy22z/community.lexicon.alpha.actor.profile/app.bsky)
- [Smoke Signal profile (using planned community.lexicon.calendar)](https://pdsls.dev/at/did:plc:erqwpimsmwnzohwoiojsy22z/community.lexicon.alpha.actor.profile/community.lexicon.calendar)
- [Picosky profile](https://pdsls.dev/at/did:plc:erqwpimsmwnzohwoiojsy22z/community.lexicon.alpha.actor.profile/social.psky)

**Reply by ghost** [reply]
Note for posterity: I had decided to switch from the `generic` record key to `com.atproto` to fit the style of the other proposed rkeys.

**Comment by ngerakines** [comment]
Account profiles are an interesting use case. They represent common elements like display name, description, and avatar, but how and where they are used is very specific to the applications that create them. They also don't imply that information should be shared or used outside of the application. I suspect that my frontpage.fyi profile would be different from my bsky.app profile, and same for smokesignal.events.

I think it'd be useful to have someone from Bluesky Social to weigh in and give some context around how they landed "self" in `at://DID/collection/self` instead of a record type reference.

**Reply by ghost** [reply]
The intended behavior is that the `self`/`generic`/`com.atproto` record key (the choice still up for debate) contains the "base" profile that services display when their own profile record is not defined. This should let users have different profiles for any service. The [examples I posted](https://github.com/orgs/lexicon-community/discussions/9#discussioncomment-11431028) are meant to illustrate this; my bsky.app profile is using (the values for `avatar`, `banner`, etc.), overriding (the `displayName`), *and* extending my base profile (with a Bluesky-specific `pinnedPost`).

**Comment by essential-randomness** [comment]
Hello! I had a chat with folks in my community about this proposal, and here's some thoughts we have. You can see the live discussion [in this video](https://www.twitch.tv/videos/2326875972?t=00h37m00s). This is our first time contributing to an effort like this, and we hope this feedback matches what you're looking for 🤞

1. The user description being 256 characters feels like a very "twitter-like" limitation. It's too long for certain use cases (IIRC from the discussion both Discord and GitHub), and too short for others (e.g. DreamWidth and AO3 both allow much longer descriptions). We discussed the possibility of having "small/medium/large" descriptions a service could choose from, according to where (and how) they want to use it. This could be achieved by making the description field be an array, so that multiple descriptions of different sizes are possible, potentially defining pre-determined sizes with different limits.
2. Similarly, the banner image is displayed very differently across different services, and it's hard to create a single one that works across many different aspect ratios. Ideally, a user would be able to add different banner images that the service can choose from, with different WxH so that the platform could choose the one best suited.
3. Regardless of multiple banner image sizes, it'd be nice to have a concept of where the focal point of the image is for when it needs to be "partially cut" (similar to `object-position` in CSS).
4. It'd be nice to add "alt-text" to images from the get go. 
5. We were wondering about the limitation of only accepting png/jpg for banner and avatar? The people demand (optional) animated avatars :P
6. Having timezone as part of the smoke signal lexicon makes sense as an example, but a user would probably want to set it across services, so that feels like a candidate for existing in the profile community lexicon and have it overridden per service as needed.
7. It's nice to have facets in descriptions, and for URLs and mentions that feels intuitive, but we spent some time wondering about how the "tag" facet should be interpreted across different services if one was present in the description? Not sure we have a solution here, but it's probably a good discussion to have.

...and this is all for now! Thanks for this initiative, very happy to see it! We'll also be looking at the social graph proposal soon :D

**Reply by ghost** [reply]
1. The limitation on description length can be overridden by a service's profile Lexicon. I think 256 is fine for the "generic" profile, but I'd support increasing it to 300.
2. The user should be able to set different banners for different services, and thus adapt their banner to whatever aspect ratio necessary for the service.
3. I support adding this, but I think it should also be up to the service's app to allow the user to crop their uploaded banner and choose the focal point *before* upload.
4. I support alt-text for avatars and banners.
5. Should be easy, just replacing the accepted types with `["image/*"]`.
6. I support putting a timezone property in the "generic" profile.

Thank you for your response, I'll be considering all this when drafting the PR soon!

**Reply by essential-randomness** [reply]
Thanks for the reply!

My main concern with having only a single description of even 300 characters and only a single banner, is that it overall forces most implementation of services based on AtProto to have a similar shape (unless they want additional complexity), which artificially limits the ability to create different experiences. It also complicates my life as a user. I believe that being too prescriptive here is going to not only make things harder for future extensibility (which I'm fairly confident will emerge as a need regardless of current choice), but also do me a disservice as a member of the ecosystem.

Let me make an example that I hope illustrates.

Imagine that I'm a fanworks fandom user, and the sites that I use most of all in the ATMosphere are "reimplementations" of two of my favorite sites: AO3 and Dreamwidth. Both of these sites allow (by design) very rich profile description experiences. Now, I go signup for those sites and given that the short description I have in my profile isn't enough for expressing myself in these spaces I go and write up a new description on both sites, which ends up in their respective Lexicons. This has already impacted my experience because, since these sites have made the choice to give me a different (and still valid!) experience, I had to go and put the same description in two different places. 

Now, I want to go and change my description. Because this information is saved in different Lexicons, instead of doing so in a single place like I could with the 300 characters description in my profile, I must go and remember all the sites that accept a longer description, maybe in a slightly different format (cause they had nothing to standardize on), and input it manually. Once again, my experience is impacted because of an "arbitrary" choice.

With a system that allowed me to set descriptions of different lengths, I could have edited my "long-form description" and have it propagated. This is what people whose "everyday sites" are 300-characters-shaped experiences
get when they join the ATMosphere! But because that isn't the web I live in and how my community expresses itself online, now I must remember and edit every site that allows me more freedom of wordiness on my profile, which ends up in a really poor experience for me (and the sites). As a result, since people don't want to bother with this additional complexity, most sites end up defaulting to only allowing 300 characters, which ends up artificially favoring the "twitter shape" in the ATMosphere ecosystem.

((Edit: I can make a similar argument as these for the banners. As someone who currently maintains 5 different social accounts on 4 different social sites and often has to change them for specific initiatives or because of sales and events "you can go to each site and redo the whole process so you can adapt it to how the site wants it to look" feels like a non-solution that will once again artificially constraint the shape of ATMosphere sites.))

**Reply by jhuckabee** [reply]
@essential-randomness - Your points about not constraining the ecosystem to a single format really resonated. Would approaching this with predefined formats instead of arbitrary lengths address your concerns? e.g., for narratives support, ~3 different formats:

* compact (300 chars) - for Twitter-like experiences
* standard (1000 chars) - for richer profiles
* extended (5000 chars) - for blog-like platforms

and for banners, maybe something like:

* compact (16:9)
* wide (21:9)
* square (1:1)

Services could then choose which formats they support, and users could provide content in multiple formats. Something like...

```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "description": "A declaration of an AT Protocol account profile with support for multiple content formats",
  "defs": {
    "main": {
      "type": "record",
      "record": {
        "type": "object",
        "properties": {
          "displayName": {
            "type": "string",
            "maxGraphemes": 64,
            "maxLength": 640
          },
          "narratives": {
            "type": "array",
            "items": { 
              "type": "ref",
              "ref": "#narrativeContent"
            },
            "maxLength": 3
          },
          "banners": {
            "type": "array",
            "items": {
              "type": "ref",
              "ref": "#bannerImage" 
            },
            "maxLength": 3
          },
          "avatar": {
            "type": "blob",
            "accept": ["image/*"],
            "maxSize": 1000000
          }
        }
      }
    },
    "narrativeContent": {
      "type": "object",
      "properties": {
        "format": {
          "type": "string",
          "enum": ["compact", "standard", "extended"]
        },
        "text": {
          "type": "string",
          "description": "Profile description text in the specified format",
          "maxGraphemes": {
            "$ref": "#formatLengths"
          }
        },
        "facets": {
          "type": "array",
          "items": {
            "type": "ref",
            "ref": "app.bsky.richtext.facet"
          }
        }
      }
    },
    "bannerImage": {
      "type": "object",
      "properties": {
        "format": {
          "type": "string",
          "enum": ["compact", "wide", "square"],
          "description": "Banner format - compact (16:9), wide (21:9), square (1:1)"
        },
        "image": {
          "type": "blob",
          "accept": ["image/*"],
          "maxSize": 2000000
        },
        "alt": {
          "type": "string",
          "maxLength": 300
        },
        "focalPoint": {
          "type": "object",
          "properties": {
            "x": {
              "type": "number",
              "minimum": 0,
              "maximum": 1
            },
            "y": {
              "type": "number", 
              "minimum": 0,
              "maximum": 1
            }
          }
        }
      }
    },
    "formatLengths": {
      "type": "integer",
      "default": 300,
      "enum": [
        300,   // compact
        1000,  // standard  
        5000   // extended
      ]
    }
  }
}
```

**Reply by essential-randomness** [reply]
> Your points about not constraining the ecosystem to a single format really resonated.

Thank you so much :) I often worry most developers won't care as much about this as they tend to be better served by restrictive formats than my community.

> Would approaching this with predefined formats instead of arbitrary lengths address your concerns? e.g., for narratives support, ~3 different formats:

I think so! In general, making this an array rather than a single field would already be more future proof and allow to create multiple descriptions that can be reused across sites. The other point of contention beyond length would be what type of formatting will be allowed. We already have a discussion on the difficulties there in #22!

I have mixed feelings on pre-determined lengths/ratios, but I think in general that would be a good thing to provide. There should probably be support for arbitrary one, but they could serve as hint for "these are generally what we suggest to pre-fill in the ecosystem".

**Reply by snarfed** [reply]
Image aspect ratio, resolution, etc are interesting! And definitely not limited to profile pictures and banners. I wonder if we should split that out into a separate lexicon that profiles, posts, etc could eventually reuse?

Also, HTML's `<img srcset>` would be good prior art to look at, at least for the schema and features, if (obviously) not the syntax. https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset

**Comment by yamarten** [comment]
Please note that due to protocol specifications, multiple schemas cannot be directly linked to one collection. The `literal` key type does not accept other keys, so you can only choose one of the examples.

FYI, followings are some alternatives that I can think of.

<details><summary>Details</summary>
<p>

### Refer to the same information from multiple services

Only generic information (base profile) is defined, and other information is defined in the collection for each service.

```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "key": "literal:self",
      "record": {
        "type": "object",
        "properties": {
          "displayName": { "type": "string" }
        }
      }
    }
  }
}
```

The service can be explicitly referenced from the record or automatically mixed in with the API.

```json
{
  "lexicon": 1,
  "id": "example.service.profile",
  "defs": {
    "main": {
      "type": "record",
      "key": "literal:self",
      "record": {
        "type": "object",
        "properties": {
          "importFrom" : {
            "type": "string",
            "format": "at-uri"
          },
          "serviceInfo": { "type": "string" }
        }
      }
    }
  }
}
```

There is also a way to separate records by field to show partial references.


### Commonize the structure of the profile

Allow the content to be defferent for each service, and if you want to unify only the type, define the type for the field, not the record.

```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "object",
      "properties": {
        "displayName": { "type": "string" }
      }
    }
  }
}
```

The profile record of each service refers to this. 

```json
{
  "lexicon": 1,
  "id": "example.service.profile",
  "defs": {
    "main": {
      "type": "record",
      "key": "literal:self",
      "record": {
        "type": "object",
        "properties": {
          "genericProfile" : {
            "type": "ref",
            "ref": "community.lexicon.actor.profile"
          },
          "serviceInfo": { "type": "string" }
        }
      }
    }
  }
}
```

### Collect profiles in one collection

In contrast to the previous one, refer to the object definition of each service. Please note that `body` is an open union, so any external object types are possible.

```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "key": "nsid",
      "record": {
        "type": "object",
        "required": ["body"],
        "properties": {
          "body": { "type": "union", "refs": ["#generic"] }
        }
      }
    },
    "generic": {
      "type": "object",
      "properties": {
        "displayName": { "type": "string" }
      }
    }
  }
}
```

The service defines the object to be referenced to this record and creates an `community.lexicon.actor.profile` record containing it.

```json
{
  "lexicon": 1,
  "id": "example.service.profile",
  "defs": {
    "main": {
      "type": "object",
      "properties": {
        "generic" : {
          "type": "string",
          "format": "nsid",
          "default": "community.lexicon.actor.profile"
        },
        "serviceInfo": { "type": "string" }
      }
    }
  }
}
```

### List the profile

If you want to make a collection to make it easier to discover profiles for other services, simply collecting references may be enough.

```json
{
  "lexicon": 1,
  "id": "community.lexicon.actor.profile",
  "defs": {
    "main": {
      "type": "record",
      "key": "literal:self",
      "record": {
        "type": "object",
        "properties": {
          "profiles": {
            "type": "array",
            "items": { "type": "string", "format": "at-uri" }
          }
        }
      }
    }
  }
}
```

There are few uses for such references, but they may be meaningful for services that refer to records from any service. For example, AT Profile has a field that declares the collection to be imported, which is similar to this.

</p>
</details>

**Reply by yamarten** [reply]
In my opinion, while other lexicon.community proposals provide useful options for developers, this discussion is trying to build a more universal mechanism. This is a respectable job, but as already discussed, it will be very difficult to find a universal profile schema.

It's hard for me to discuss the schemas that should be, so this is just a list of possible options on the Lexicon system rather than a recommendation.
